Add timeout for safe browsing check

pascoej opened the pull request:
#### 1241768fd694be6850ad1fe896566206c31e47aa
<pre>
Add timeout for safe browsing check
<a href="https://bugs.webkit.org/show_bug.cgi?id=287851">https://bugs.webkit.org/show_bug.cgi?id=287851</a>
<a href="https://rdar.apple.com/142507598">rdar://142507598</a>

Reviewed by Alex Christensen.

In the case where the safe browsing daemon hangs on our request, we
should allow requests to go through after waiting 2x the amount of time
the response took. This implements such behavior and adds API tests for it.

* Source/WebKit/Platform/spi/Cocoa/SafeBrowsingSPI.h:
* Source/WebKit/UIProcess/API/APINavigation.cpp:
(API::Navigation::resetRequestStart):
* Source/WebKit/UIProcess/API/APINavigation.h:
(API::Navigation::setSafeBrowsingCheckOngoing):
(API::Navigation::safeBrowsingCheckOngoing const):
(API::Navigation::setSafeBrowsingWarning):
(API::Navigation::safeBrowsingWarning const):
(API::Navigation::requestStart const):
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::beginSafeBrowsingCheck):
* Source/WebKit/UIProcess/WebFramePolicyListenerProxy.cpp:
(WebKit::WebFramePolicyListenerProxy::didReceiveSafeBrowsingResults):
* Source/WebKit/UIProcess/WebFramePolicyListenerProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didReceiveServerRedirectForProvisionalLoadForFrameShared):
(WebKit::WebPageProxy::beginSafeBrowsingCheck):
(WebKit::WebPageProxy::decidePolicyForNavigationAction):
(WebKit::WebPageProxy::decidePolicyForResponseShared):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SafeBrowsing.mm:
(+[DelayedLookupContext sharedLookupContext]):
(-[DelayedLookupContext lookUpURL:completionHandler:]):
(TEST(SafeBrowsing, HangTimeout)):
(TEST(SafeBrowsing, PostResponse)):
(TEST(SafeBrowsing, PreresponseSafeBrowsingWarning)):

Canonical link: <a href="https://commits.webkit.org/294690@main">https://commits.webkit.org/294690@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cc99b1f238f560f18c673db7d3d106296ad3006b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/102712 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/22385 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/12704 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/107877 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/53353 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/104751 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/22696 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/30887 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/78116 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35083 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/105718 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/17585 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/92695 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58448 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/17426 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/10729 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/52710 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/87241 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10798 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/110253 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/29849 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/22004 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87099 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/30213 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/88888 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/86705 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22076 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31536 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9267 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/24102 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/29776 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/35096 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/29584 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/32911 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/31146 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/WebFramePolicyListenerProxy.h:
```diff
@@ -71,6 +76,8 @@ class WebFramePolicyListenerProxy : public API::ObjectImpl<API::Object::Type::Fr
     std::optional<std::optional<NavigatingToAppBoundDomain>> m_isNavigatingToAppBoundDomain;
     bool m_doneWaitingForLinkDecorationFilteringData { false };
+    bool m_doneWaitingForSafeBrowsing { false };
+    WebCore::Timer m_safeBrowsingTimeout;
```
We cannot use `WebCore::Timer` in the UI process. Use `RunLoop::Timer`.

achristensen07 commented on Source/WebKit/UIProcess/WebFramePolicyListenerProxy.h:
```diff
@@ -45,6 +47,8 @@ enum class ShouldExpectAppBoundDomainResult : bool { No, Yes };
 enum class ShouldWaitForInitialLinkDecorationFilteringData : bool { No, Yes };
 enum class WasNavigationIntercepted : bool { No, Yes };
 
+static constexpr Seconds defaultSafeBrowsingTimeout { 2_s };
```
I think this is not enough for many network configurations.  The time should be based on how much time the request roundtrip to the server took.

achristensen07 commented on Source/WebKit/UIProcess/WebFramePolicyListenerProxy.h:
```diff
@@ -71,6 +76,8 @@ class WebFramePolicyListenerProxy : public API::ObjectImpl<API::Object::Type::Fr
     std::optional<std::optional<NavigatingToAppBoundDomain>> m_isNavigatingToAppBoundDomain;
     bool m_doneWaitingForLinkDecorationFilteringData { false };
+    bool m_doneWaitingForSafeBrowsing { false };
+    RunLoop::Timer m_safeBrowsingTimeout;
```
This seems like it's not really the responsibility of the listener, but more the responsibility of the WebPageProxy.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -8183,6 +8184,9 @@ void WebPageProxy::decidePolicyForResponseShared(Ref<WebProcessProxy>&& process,
         m_openedMainFrameName = { };
     }
 
+    auto elapsed = (WallTime::now() - m_pageLoadTiming->navigationStart());
```
WebPageLoadTiming's m_navigationStart is never set.  You'll have to do that if you want to use this value.

pascoej replied:
It's set in the constructor of WebPageLoadTiming

achristensen07 commented on Source/WebKit/UIProcess/WebFramePolicyListenerProxy.cpp:
```diff
@@ -62,6 +62,7 @@ void WebFramePolicyListenerProxy::didReceiveAppBoundDomainResult(std::optional<N
 void WebFramePolicyListenerProxy::didReceiveSafeBrowsingResults(RefPtr<BrowsingWarning>&& safeBrowsingWarning)
 {
     ASSERT(isMainRunLoop());
+
```
nit: don't touch this file just to add whitespace.

achristensen07 commented on Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
```diff
@@ -243,8 +243,25 @@ static bool exceedsRenderTreeSizeSizeThreshold(uint64_t thresholdSize, uint64_t
+                return;
+
+            if (protectedThis->internals().timeoutSafeBrowsingCheckTimer && protectedThis->internals().timeoutSafeBrowsingCheckTimer->isActive())
+                protectedThis->internals().timeoutSafeBrowsingCheckTimer->stop();
```
This isn't necessarily for the same navigation.

achristensen07 commented on Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
```diff
@@ -243,8 +243,25 @@ static bool exceedsRenderTreeSizeSizeThreshold(uint64_t thresholdSize, uint64_t
-        RunLoop::protectedMain()->dispatch([listener = WTFMove(listener), result = retainPtr(result), error = retainPtr(error), forMainFrameNavigation, url = WTFMove(url)] {
+
+    // This gets adjusted whenever we get the response.
+    auto initialTimeout = Seconds(5_s);
```
We don't want this.  Just start the timer when we receive the response if it hasn't received SSB results already.

achristensen07 commented on Source/WebKit/UIProcess/Cocoa/BrowsingWarningCocoa.mm:
```diff
@@ -71,7 +71,6 @@ static String localizedProviderShortName(SSBServiceLookupResult *result)
         return "Tencent"_s;
     if ([result.provider isEqual:SSBProviderApple])
         return "Apple"_s;
-    ASSERT_NOT_REACHED();
```
I think we want to leave this.  I changed the test provider name instead.

achristensen07 commented on Source/WebKit/UIProcess/API/APINavigation.cpp:
```diff
@@ -51,6 +51,7 @@ Navigation::Navigation(WebCore::ProcessIdentifier processID)
     : m_navigationID(WebCore::NavigationIdentifier::generate())
     , m_processID(processID)
     , m_clientNavigationActivity(ProcessThrottler::TimedActivity::create(navigationActivityTimeout))
+    , m_navigationStart(MonotonicTime::now())
```
Let's call this m_requestStart and reset it when an HTTP redirect happens.

achristensen07 commented on Source/WebKit/UIProcess/API/APINavigation.h:
```diff
@@ -208,6 +216,10 @@ class Navigation : public ObjectImpl<Object::Type::Navigation> {
     std::optional<OptionSet<WebCore::AdvancedPrivacyProtections>> m_originatorAdvancedPrivacyProtections;
+    RefPtr<RunLoop::DispatchTimer> m_timeoutSafeBrowsingCheckTimer { nullptr };
+    MonotonicTime m_navigationStart;
+    bool m_hasReceivedSafeBrowsingResult { false };
```
This state should be in the listener.

achristensen07 commented on Source/WebKit/UIProcess/API/APINavigation.h:
```diff
@@ -208,6 +216,10 @@ class Navigation : public ObjectImpl<Object::Type::Navigation> {
     bool m_isLoadedWithNavigationShared { false };
     RefPtr<API::WebsitePolicies> m_websitePolicies;
     std::optional<OptionSet<WebCore::AdvancedPrivacyProtections>> m_originatorAdvancedPrivacyProtections;
+    RefPtr<RunLoop::DispatchTimer> m_timeoutSafeBrowsingCheckTimer { nullptr };
```
{ nullptr } is unnecessary.

achristensen07 commented on Source/WebKit/UIProcess/WebFramePolicyListenerProxy.h:
```diff
@@ -68,7 +68,6 @@ class WebFramePolicyListenerProxy : public API::ObjectImpl<API::Object::Type::Fr
     WebFramePolicyListenerProxy(Reply&&, ShouldExpectSafeBrowsingResult, ShouldExpectAppBoundDomainResult, ShouldWaitForInitialLinkDecorationFilteringData);
 
     std::optional<std::pair<RefPtr<API::WebsitePolicies>, ProcessSwapRequestedByClient>> m_policyResult;
-    std::optional<RefPtr<BrowsingWarning>> m_safeBrowsingWarning;
```
I think this should stay here.

achristensen07 commented on Source/WebKit/UIProcess/WebFramePolicyListenerProxy.cpp:
```diff
@@ -62,7 +62,6 @@ void WebFramePolicyListenerProxy::didReceiveAppBoundDomainResult(std::optional<N
 void WebFramePolicyListenerProxy::didReceiveSafeBrowsingResults(RefPtr<BrowsingWarning>&& safeBrowsingWarning)
```
This should probably add an early return if we've already received safe browsing results for this listener.  Otherwise the second call might confuse the state.

achristensen07 commented on Source/WebKit/UIProcess/WebFrameProxy.h:
```diff
@@ -214,6 +214,7 @@ class WebFrameProxy : public API::ObjectImpl<API::Object::Type::Frame>, public C
     void updateOpener(WebCore::FrameIdentifier);
     WebFrameProxy* opener() { return m_opener.get(); }
     void disownOpener() { m_opener = nullptr; }
+    RefPtr<WebFramePolicyListenerProxy> activeListener() { return m_activeListener; }
```
This is never set.

pascoej replied:
It is set in `WebFrameProxy::setUpPolicyListenerProxy`

achristensen07 commented on Source/WebKit/UIProcess/API/APINavigation.h:
```diff
@@ -175,8 +176,13 @@ class Navigation : public ObjectImpl<Object::Type::Navigation> {
     API::WebsitePolicies* websitePolicies() { return m_websitePolicies.get(); }
     RefPtr<API::WebsitePolicies> protectedWebsitePolicies() const { return m_websitePolicies; }
 
+
```
Don't add this newline.

achristensen07 commented on Source/WebKit/UIProcess/API/APINavigation.cpp:
```diff
@@ -82,6 +85,7 @@ Navigation::Navigation(WebCore::ProcessIdentifier processID, Ref<WebBackForwardL
     , m_fromItem(WTFMove(fromItem))
     , m_backForwardFrameLoadType(backForwardFrameLoadType)
     , m_clientNavigationActivity(ProcessThrottler::TimedActivity::create(navigationActivityTimeout))
+    , m_requestStart(MonotonicTime::now())
```
All these can be an initializer in the header instead of adding to each constructor.

achristensen07 commented on Source/WebKit/Platform/spi/Cocoa/SafeBrowsingSPI.h:
```diff
@@ -70,9 +71,12 @@ WTF_EXTERN_C_END
 + (SSBLookupContext *)sharedLookupContext;
 
 - (void)lookUpURL:(NSURL *)URL completionHandler:(void (^)(SSBLookupResult *, NSError *))completionHandler;
+- (void)lookUpURL:(NSURL *)URL isMainFrame:(bool)isMainFrame hasHighConfidenceOfSafety:(BOOL)hasHighConfidenceOfSafety completionHandler:(void (^)(SSBLookupResult *, NSError * _Nullable))completionHandler;
```
I think we need a test that swizzles this and responds immediately with a phishing result and verify it shows a warning and with a nothing result and verify it loads without showing a warning.

pascoej replied:
I think those two cases are already covered by `SafeBrowsing WKWebViewGoBack`

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SafeBrowsing.mm:
```diff
@@ -536,4 +541,98 @@ + (NullLookupContext *)sharedLookupContext
+        RunLoop::protectedMain()->dispatchAfter(1000_s, [task = retainPtr(task)] {
+            auto response = adoptNS([[NSURLResponse alloc] initWithURL:task.get().request.URL MIMEType:@"text/html" expectedContentLength:0 textEncodingName:nil]);
+            [task didReceiveResponse:response.get()];
+            [task didReceiveData:[NSData dataWithBytes:mainResource length:strlen(mainResource)]];
```
NSData.data

achristensen07 commented on Source/WebKit/UIProcess/WebFramePolicyListenerProxy.h:
```diff
@@ -45,7 +45,7 @@ enum class ShouldExpectAppBoundDomainResult : bool { No, Yes };
 enum class WasNavigationIntercepted : bool { No, Yes };
 
-class WebFramePolicyListenerProxy : public API::ObjectImpl<API::Object::Type::FramePolicyListener> {
+class WebFramePolicyListenerProxy : public API::ObjectImpl<API::Object::Type::FramePolicyListener>, public CanMakeWeakPtr<WebFramePolicyListenerProxy>  {
```
nit: extra space

achristensen07 commented on Source/WebKit/UIProcess/API/APINavigation.h:
```diff
@@ -27,6 +27,7 @@
 
 #include "APIObject.h"
 #include "APIWebsitePolicies.h"
+#include "BrowsingWarning.h"
```
Can this be forward declared instead?

pascoej replied:
I don't think so (because it's used as the inner type of a RefPtr<T>.)
`.../OpenSource/WebKitBuild/Debug/usr/local/include/wtf/Ref.h:62:16: error: member access into incomplete type 'WebKit::BrowsingWarning'
   62 |             ptr->deref();`

achristensen07 replied:
It can if you make setSafeBrowsingWarning and safeBrowsingWarning no longer inline.

achristensen07 commented on Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
```diff
@@ -228,32 +229,37 @@ static bool exceedsRenderTreeSizeSizeThreshold(uint64_t thresholdSize, uint64_t
+    if (navigation)
+        navigation->setSafeBrowsingCheckOngoing(true);
+
+    auto completionHandler = makeBlockPtr([navigation = WTFMove(navigation), forMainFrameNavigation = frame.isMainFrame(), url = url, weakThis = WeakPtr { *this }, frame = Ref { frame }] (SSBLookupResult *result, NSError *error) mutable {
```
nit: " = url" is unnecessary.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -6755,8 +6755,10 @@ void WebPageProxy::didReceiveServerRedirectForProvisionalLoadForFrameShared(Ref<
-    if (navigation)
+    if (navigation) {
         navigation->appendRedirectionURL(request.url());
+        navigation->resetRequestStart();
```
I'm not sure we actually need this.  If a redirect happens for the main frame main resource, we get another decidePolicyForNavigationAction I think.  Could you add a test that redirects the main frame main resource?

pascoej replied:
We get another decidePolicyForNavigationAction, but it's the same navigation, so resetting the start time here ensures we only later wait for the period of the last redirect's request.

I'm changing the way ongoing requests for navigations are handled because there might be multiple.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -7890,8 +7891,8 @@ void WebPageProxy::decidePolicyForNavigationAction(Ref<WebProcessProxy>&& proces
-    ] (PolicyAction policyAction, API::WebsitePolicies* policies, ProcessSwapRequestedByClient processSwapRequestedByClient, RefPtr<BrowsingWarning>&& safeBrowsingWarning, std::optional<NavigatingToAppBoundDomain> isAppBoundDomain, WasNavigationIntercepted wasNavigationIntercepted) mutable {
-        WEBPAGEPROXY_RELEASE_LOG(Loading, "decidePolicyForNavigationAction: listener called: frameID=%" PRIu64 ", isMainFrame=%d, navigationID=%" PRIu64  ", policyAction=%" PUBLIC_LOG_STRING ", safeBrowsingWarning=%d, isAppBoundDomain=%d, wasNavigationIntercepted=%d", frame->frameID().toUInt64(), frame->isMainFrame(), navigation ? navigation->navigationID().toUInt64() : 0, toString(policyAction). ...
+    ] (PolicyAction policyAction, API::WebsitePolicies* policies, ProcessSwapRequestedByClient processSwapRequestedByClient, RefPtr<BrowsingWarning>&&, std::optional<NavigatingToAppBoundDomain> isAppBoundDomain, WasNavigationIntercepted wasNavigationIntercepted) mutable {
+        WEBPAGEPROXY_RELEASE_LOG(Loading, "decidePolicyForNavigationAction: listener called: frameID=%" PRIu64 ", isMainFrame=%d, navigationID=%" PRIu64  ", policyAction=%" PUBLIC_LOG_STRING ", safeBrowsingWarning=%d, isAppBoundDomain=%d, wasNavigationIntercepted=%d", frame->frameID().toUInt64(), frame->isMainFrame(), navigation ? navigation->navigationID().toUInt64() : 0, toString(policyAction). ...
```
If safeBrowsingWarning=0 is always printed, we should consider removing it.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -8193,16 +8203,73 @@ void WebPageProxy::decidePolicyForResponseShared(Ref<WebProcessProxy>&& process,
+                auto transaction = protectedPageLoadState->transaction();
+                protectedPageLoadState->setTitleFromBrowsingWarning(transaction, { });
+
+                switchOn(result, [&protectedThis, &completionHandler] (const URL& url) {
```
I would just capture &.  Same below.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -8173,10 +8176,17 @@ void WebPageProxy::decidePolicyForResponseShared(Ref<WebProcessProxy>&& process,
+        requestStart = navigation->requestStart();
+    }
+
+    Ref listener = frame->setUpPolicyListenerProxy([this, protectedThis = Ref { *this }, completionHandler = WTFMove(completionHandler), frameInfo = WTFMove(frameInfo), navigation, process, navigationResponse, request, frame] (PolicyAction policyAction, API::WebsitePolicies*, ProcessSwapRequestedByClient processSwapRequestedByClient, RefPtr<BrowsingWarning>&&, std::optional<NavigatingToAppBoundDo ...
```
There's a lot of things captured.  Could you put them each on their own line like we do elsewhere in this area?

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -8193,16 +8203,73 @@ void WebPageProxy::decidePolicyForResponseShared(Ref<WebProcessProxy>&& process,
+                protectedPageLoadState->commitChanges();
+            }
+
+            if (!frame->isMainFrame()) {
```
There aren't any tests with iframes that hit this.

achristensen07 commented on Source/WebKit/UIProcess/API/APINavigation.cpp:
```diff
@@ -119,8 +124,8 @@ void Navigation::appendRedirectionURL(const WTF::URL& url)
     return currentRequestIsRedirect()
-        && m_lastNavigationAction
-        && RegistrableDomain(m_lastNavigationAction->redirectResponse.url()) != RegistrableDomain(m_currentRequest.url());
+    && m_lastNavigationAction
```
This indentation change should be undone.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -8191,16 +8210,73 @@ void WebPageProxy::decidePolicyForResponseShared(Ref<WebProcessProxy>&& process,
-        if (policyAction == PolicyAction::Use && PreviewConverter::supportsMIMEType(navigationResponse->response().mimeType())) {
-            auto auditToken = process->connection().getAuditToken();
+        if (policyAction == PolicyAction::Use && supportsMIMEType) {
+            auto auditToken = m_legacyMainFrameProcess->connection().getAuditToken();
```
Why are we switching from the process from which Messages::WebPageProxy::DecidePolicyForResponse came to the main frame process?  That seems like a step in the wrong direction when site isolation is enabled.

achristensen07 commented on Source/WebKit/UIProcess/API/APINavigation.h:
```diff
@@ -211,6 +222,9 @@ class Navigation : public ObjectImpl<Object::Type::Navigation> {
     std::optional<OptionSet<WebCore::AdvancedPrivacyProtections>> m_originatorAdvancedPrivacyProtections;
+    MonotonicTime m_requestStart { MonotonicTime::now() };
+    RefPtr<WebKit::BrowsingWarning> m_safeBrowsingWarning;
+    ListHashSet<size_t> m_ongoingSafeBrowsingChecks;
```
I feel like we may run into issues putting 0 into this ListHashSet, but there are no checks that the size_t is nonzero.

pascoej replied:
I've confirmed inserting and removing 0 works.

achristensen07 reviewed (commented):
It seems like you added some logic having to do with redirect chains, but you have no tests with redirect chains.  Do we check all the URLs in the redirect chain and only proceed if they are all safe, or do we only check the last one?  It seems like the latter would be more performant, but the former would be safer.

achristensen07 commented on Source/WebKit/UIProcess/API/APINavigation.cpp:
```diff
@@ -142,6 +148,44 @@ void Navigation::markRequestAsFromClientInput()
+{
+    m_safeBrowsingWarning = WTFMove(safeBrowsingWarning);
+}
+
```
nit: extra newline.

achristensen07 commented on Source/WebKit/UIProcess/API/APINavigation.cpp:
```diff
@@ -142,6 +148,44 @@ void Navigation::markRequestAsFromClientInput()
+        index = m_redirectChain.size();
+    return index;
+}
+
```
ditto

achristensen07 commented on Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
```diff
@@ -228,32 +229,38 @@ static bool exceedsRenderTreeSizeSizeThreshold(uint64_t thresholdSize, uint64_t
+    SSBLookupContext *context = [SSBLookupContext sharedLookupContext];
+    if (!url.isValid() || !context)
+        return;
+    size_t redirectChainIndex = navigation->redirectChainLengthIndex(url);
```
Let's remove "Length" from the name of this function.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -7938,7 +7939,9 @@ void WebPageProxy::decidePolicyForNavigationAction(Ref<WebProcessProxy>&& proces
         protectedPageClient->clearBrowsingWarning();
 
+        RefPtr safeBrowsingWarning = navigation->safeBrowsingWarning();
         if (safeBrowsingWarning) {
```
Can this be if (RefPtr safeBrowsingWarning = ...)?

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SafeBrowsing.mm:
```diff
@@ -536,4 +541,222 @@ + (NullLookupContext *)sharedLookupContext
+    while (![webView _safeBrowsingWarning])
+        TestWebKitAPI::Util::spinRunLoop();
+}
+
```
nit: extra newline.

pascoej committed:
Add timeout for safe browsing check
https://bugs.webkit.org/show_bug.cgi?id=287851
rdar://142507598

Reviewed by Alex Christensen.

In the case where the safe browsing daemon hangs on our request, we
should allow requests to go through after waiting 2x the amount of time
the response took. This implements such behavior and adds API tests for it.

* Source/WebKit/Platform/spi/Cocoa/SafeBrowsingSPI.h:
* Source/WebKit/UIProcess/API/APINavigation.cpp:
(API::Navigation::resetRequestStart):
* Source/WebKit/UIProcess/API/APINavigation.h:
(API::Navigation::setSafeBrowsingCheckOngoing):
(API::Navigation::safeBrowsingCheckOngoing const):
(API::Navigation::setSafeBrowsingWarning):
(API::Navigation::safeBrowsingWarning const):
(API::Navigation::requestStart const):
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::beginSafeBrowsingCheck):
* Source/WebKit/UIProcess/WebFramePolicyListenerProxy.cpp:
(WebKit::WebFramePolicyListenerProxy::didReceiveSafeBrowsingResults):
* Source/WebKit/UIProcess/WebFramePolicyListenerProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didReceiveServerRedirectForProvisionalLoadForFrameShared):
(WebKit::WebPageProxy::beginSafeBrowsingCheck):
(WebKit::WebPageProxy::decidePolicyForNavigationAction):
(WebKit::WebPageProxy::decidePolicyForResponseShared):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SafeBrowsing.mm:
(+[DelayedLookupContext sharedLookupContext]):
(-[DelayedLookupContext lookUpURL:completionHandler:]):
(TEST(SafeBrowsing, HangTimeout)):
(TEST(SafeBrowsing, PostResponse)):
(TEST(SafeBrowsing, PreresponseSafeBrowsingWarning)):

Canonical link: https://commits.webkit.org/294690@main
